### ISSUE TITLE
FRB Editor should be able to translate Gum entities into native FRB entities

### ISSUE THREAD

**Issue by profexorgeek** [vchelaru/FlatRedBall#1317]
There is a test project here but it is not open source, we can consider open sourcing a version of it as a demo project but for now it is just for testing this feature and I don't want any art or features I add to be open source for now.
https://github.com/profexorgeek/Nardiak/

![simple-rigging](https://github.com/vchelaru/FlatRedBall/assets/711100/9ca85710-1dc9-4916-9ad5-632a267afc91)

**Overall Scenario**

The goal of this feature is to allow users to create robust state-based animations for FRB entities and generally create entities in Gum, but use the entities in FRB with native components and the native rendering system.

1. User wants to create a series of entities that use interpolation-based animations defined in Gum
2. User creates placeholder FRB entities for each Gum object they want to convert
3. User creates Gum entities, giving them sprites, shapes, states, animations, events, etc
4. User adds Gum component file in FRB entity and specifies add mode is "convert" or something similar
5. User is warned that any matching named objects will have their properties overwritten ("don't show again" option on warning?)
6. FRB creates entity components for each Gum item that doesn't already exist. If names match, it replaces all properties on the FRB object with the incoming Gum object
7. Entity creation is recursive, if a Gum object references another Gum object, the user is prompted to choose which FRB entity should be populated by the child Gum object
8. User can cancel the conversion and create child objects if they have failed to do so
9. Once conversion is complete, user can add entity instance to game screen and write game logic in C# against components, events, and animations as they would any normal FRB entity
12. State animations can be played and their timing can be scaled just like frame-based animations

**Requirements**

- [ ] User should be able to import a gum component and specify that it should be converted into a native FRB object
  - [ ] Gum states should become FRB states
  - [ ] Gum sprites should become FRB sprites
  - [ ] Gum shapes should become FRB shapes
  - [ ] Gum parent/child relationships should be preserved
  - [ ] Gum animation events should become FRB events
  - [ ] Constraints on allowed coordinate systems are okay since FRB doesn't support all of the dynamic coordinate calculation that exists in Gum
- [ ] Conversions should be individually specified per entity, not bulk interpreted from Gum so developer has control over what remains native to Gum and what is converted to FRB
- [ ] Rendering of all components on the entity should be native FRB rendering, not Gum rendering
- [ ] Coding against components should work the same as coding against any normal FRB entity
- [ ] Converted entities should support FRB inheritance so logic for damage, movement, and other shared things still works
- [ ] When converting or refreshing the conversion after Gum changes, components with matching names (whether native to the entity or from inheritance) should be populated by the gum object
- [ ] Animations should be playable in code and use states
- [ ] Animation timing should be scalable in code
- [ ] Animations should be stoppable, switchable, and any instruction systems should be cleared on stop/switch
- [ ] Sprite textures should be swappable in code as normal, for example the developer might change the animation chain used by a sprite that was converted from Gum
- [ ] ~~Conversion direction is one-way, meaning all FRB changes are overwritten by Gum changes and any changes to FRB states or components are lost when the dev chooses to refresh the conversion.~~
- [ ] Custom code, such as events, is not overwritten on conversion
- [ ] ~~Refreshing the conversion is manual?~~
- [ ] When a Gum entity references other Gum entities, FRB should convert recursively and prompt you to select an FRB entity for unique Gum component
- [ ] Recursive conversion can be cancelled
- [ ] Child FRB components must be created first, dev must specify child components during recursive conversion
- [ ] Child components that have already been converted, are not reconverted if there are no changes. For example, if three different creatures are created and all have a Sword, the sword should not be recursively created for every entity if it hasn't had any changes
- [ ] Converting Gum objects entirely into C# code gen is acceptable for this iteration. As a future consideration, having FRB interpret gum objects as FRB objects in realtime would be a "nice to have"
- [ ] Gum objects can be flipped on the X axis and all animations, sprites, and shapes flip

**What the example project has**

- A gum entity called Components/Entities/Creatures/HumanoidSmall
- A placeholder FRB entity called Entities/Creatures/HumanoidSmall, which inherits from CreatureBase
- A placeholder FRB entity called Entities/Weapons/Sword, which inherits from WeaponBase
- The Gum entities use animation chains to define their sprite texture
- Looping and one-off animations in Gum
- Named events that fire in the animation
- Collision objects on HumanoidSmall that should convert
- HumanoidSmall owns an instance of Sword, which should prompt the developer to choose an entity when converting or cancel. The developer should choose the FRB entity Entities/Weapons/Sword as the conversion target for the recursive child object
- Collision objects on the child sword that should convert recursively when importing the HumanoidSmall
- A tile map with collision and nav mesh in Level1

**What the example project does NOT have**

- A Player entity
- Top down movement
- Collision relationships defined

Feel free to make a Player with top down movement that is an instance of the converted HumanoidSmall entity and plays animations when moving or pressing buttons. I didn't add the top down movement because I wanted to figure out the base structure without the complexity of an existing player and movement system.

**Comment by profexorgeek** [comment]
After discussing extensively with @vchelaru in the FRB Discord, we decided on some changes to the original requirements:

- The Gum object will be created and continue to exist on the FRB entity
- The Gum object will be the system-of-record for its component layout
- FRB will use the Gum layout engine to maintain layout concepts that native FRB doesn't support such as different origins and percentage-based positioning
- All native FRB objects will be created as stated in the requirements but child object properties such as position and rotation will be `readonly` on FRB since internal layout will still be controlled by Gum since they may use Gum-specific coordinate systems
- Editing entity-level properties such as position, rotation, velocity, etc will continue to work as expected
- If changes to internal component layout is needed, this will need to be made on the Gum object, which will propagate to the FRB object in the same frame

This was decided because it makes the first pass implementation easier and faster. It is not ideal for developers to have to work with two different coordinate systems, however we believe that working with the Gum coordinate system will be relatively rare. Developers may need to change Gum coordinates when implementing procedural animation interpolation or in cases where you want something like knockback on an individual part of an entity (such as the head responding to a punch). We believe these are advanced scenarios that will require a deeper understanding of the Gum/FRB coordinate system anyway and this is an acceptable compromise.

If the dual coordinate system causes regular confusion for developers or if maintaining the Gum layout system causes performance problems at scale, we will revisit the implementation and change or optimize.

**Comment by profexorgeek** [comment]
After more extensive discussion and testing, we are investing in support for Spine, which is much more robust and meets all of the requirements I outlined here. I'm closing this issue since we have changed course. See this issue for Spine implementation details:
https://github.com/vchelaru/FlatRedBall/issues/1320
